### DISCUSSION TITLE
Allow assigning to readonly explicit properties in the constructor

### DISCUSSION THREAD

**Discussion by RenderMichael** [dotnet/csharplang#6959] [Language Ideas] [answered]
The following code is invalid:
``` csharp
public class C : I
{
    string P { get; }

    int I.P {get;}

    public C()
    {
        ((I)this).P = 5; // error CS0200: Property or indexer 'I.P' cannot be assigned to -- it is read only
    }
}

interface I
{
    int P { get; }
}
```

I don't see a good language reason for this to be the case

edit: changed the example so the two `P`'s have different types, to make things clearer

**Comment by RenderMichael** [comment]
Every language change needs some motivation behind it, so I'll explain mine: I have a class which inherits a `Context` property from it's base class, while also implementing an interface with an unrelated read-only property, also called `Context`. My current workaround is like so: adapted for the minimal repro above:
``` csharp
public class C : I
{
    private readonly int _p;
    string P { get; }

    int I.P => _p;

    public C()
    {
        _p = 5;
    }
}

interface I
{
    int P { get; }
}
```

However, this is unpleasant, and unless there's an issue I'm not aware of, doesn't need to be the case.

edit: changed in line with the edit above

**Comment by CyrusNajmabadi** [comment, marked as answer]
How would we know what to assign to?  the interface property literally has no setter, so there's nothing for the compiler to invoke to actually run the setter code. :)

**Reply by RenderMichael** [reply]
Would this not be a auto-property where the constructor has access to the backing field? Adding a setter makes this code valid and a backing field is generated for the implicit property, as shown here [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgPgAQMwAIECYUGEUhQSQFgAoAbxJUpVgBcCA6ABRVJQHMBTGgbhQGcuvAL4kKVZNgAUASjGVyxKkpSTJ+aTQAWUPtKYoAvCgCs3OShHFLJWhwBOAMwgBjDgRIKltFM1ace/IIWJEJAA==).

Is there a problem with the constructor directly setting the backing field like with other get-only auto properties?

edit: backing field

**Reply by CyrusNajmabadi** [reply]
> Would this not be a auto-property 

No.  You're casting to `I`.  so it could be an impl of htat interface which then may not have a setter at all.  Consider if there is a derived type of C that itself reimplements `I` with a property with no setter.  There's literally nothing for teh compiler to call into.

> Is there a problem with the constructor directly setting the backing field like with other get-only auto properties?

Yes.  It would then be bypassing the interface, which could break code.

**Reply by RenderMichael** [reply]
So if I understand this correctly, the only reason there's a backing field in the example, is because the `C`'s implementation of `P` uses an auto-getter. And if a class derived from `C` re-implemented `P` with a full property, that backing field would disappear.

And since you wouldn't do anything like special-casing for sealed classes or sealing explicit properties, there's nothing that can be done here. Did I get that right?

**Reply by CyrusNajmabadi** [reply]
> And if a class derived from C re-implemented P with a full property, that backing field would disappear.

It wouldn't disappear.  But it would not be settable by anyone anymore.  

> And since you wouldn't do anything like special-casing for sealed classes

It seems hard to imagine us deciding to bypass access through an interface in any circumstance.    But we'd certainly consider any proposal and make a firm conclusion on it if presented.

**Reply by RikkiGibson** [reply]
There is a part of me that feels like, if we won't permit setting the field underlying this auto-property anywhere, we shouldn't have permitted declaring it as an auto-property. Not that we can take that back at this point.

**Reply by RikkiGibson** [reply]
One major curiosity I have here is, what happens when this is combined with primary constructors. In that case there is a clear place to assign a parameter to the auto-property. Does it end up being a gap where primary constructors can initialize these but not other kinds of constructors.

```cs
class C(int prop) : I
{
    int I.Prop { get; } = prop;
}
```

**Reply by DavidArno** [reply]
@RikkiGibson, [put the code through the primary constructors branch and it compiles just fine](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQAUEEsC2UECeAwgPYB2yMCcAxjCQsgDQAmIA1AD5ZkwQJgoNCAAIAkgFgAUAG9pIhSJ4wR6ETJEBzCDADcIgL7SjU6QAEATCKIAKZSIAOCEg4CUIkOOlypipb3EAOjUNbT1DEQBeR2cHXWMgA===)

**Comment by MichalPetryka** [comment]
I wonder if adding `readonly set` and `init set` accessors that'd be callable from the constructor would be a good idea here.

Then `I` would define:
```C#
interface I
{
    int P { get; readonly set; }
}
```
